### PR TITLE
Add and use new IntrusivePtr type in Zeek

### PR DESCRIPTION
**Pull request by Neverlord**
@timwoj we discussed the use of smart pointers for automated reference counting a while back. This PR takes the first steps towards modernizing the Zeek code base by moving away from C-style manual reference counting.

Manual memory management via Ref/Unref is verbose and prone to error. An intrusive smart pointer automates the reference counting, makes code more robust (in particular w.r.t. to exceptions) and reduces boilerplate code. A big benefit of the intrusive smart pointers for Zeek is that they can co-exist with the manual memory management. Rather than having to port the entire code base at once, we can migrate components one-by-one. In this first step, we add `ValPtr` as an alias for `intrusive_ptr<Val>` and start using it in the Broker Manager. This makes the previous `unref_guard` obsolete.

### CODE REVIEW

**Comment by rsmmr** [comment]
This seems like the right direction to me for moving towards automatic reference counting. As you say, we can introduce this incrementally; and Broker seems like a good place to start.

The main question I think is what ptr implementation to use. From what I can tell, [retain_ptr](https://api.csswg.org/bikeshed/?force=1&url=https://git.io/fxsVI) is on the way to be standardized, so I'm thinking we should use that instead of the CAF version.

Nit: I wouldn't do the typedef to `ValPtr` as that's hiding the fact that it's a new-style pointer. `retain_ptr<Val>` is more explicit, which will make it easier to tell what's going on in particular during the transition. And in combination with `auto` it shouldn't be too bad for typing either.

**Comment by Neverlord** [comment]
> The main question I think is what ptr implementation to use. From what I can tell, retain_ptr is on the way to be standardized, so I'm thinking we should use that instead of the CAF version.

I wouldn't count on that TBH. The committee rejected all intrusive types so far. This seems to be one of the many proposals that go nowhere. Even if this smart pointer would eventually make it: it didn't make the cut for C++20, so it's at least another three years away. In that time, the API for this proposal is subject to change (and most likely *will* change). So using the current draft of the implementation does not guarantee any API stability going forward.

Should this get ever voted in: I sincerely hope not under that name. C++ devs have been familiar with this pointer under the name "intrusive pointer" for ages and it has been available in the Boost library as `intrusive_ptr` for well over a decade.

In any case, no compiler ships a `retain_ptr` implementation. So Zeek would have to bundle something on its own. Is copying over another implementation really more desirable than including the CAF header that Zeek already has access to? Would you mind explaining your concerns? Is it API stability or is the general policy simply not to use CAF headers in Zeek directly?

**Comment by Neverlord** [comment]
I wanted to see why the checks fail, but I only see errors such as these:

```

curl: (7) Failed to connect to 2620:83:8000:102::4d: Cannot assign requested address

Error: Could not get https://data.zeek.org/traces/2009-M57-day11-18.trace.gz.md5sum, skipping download.

curl: (7) Failed to connect to 2620:83:8000:102::4d: Cannot assign requested address

Error: Could not get https://data.zeek.org/traces/2009-M57-day11-21.trace.gz.md5sum, skipping download.

curl: (7) Failed to connect to 2620:83:8000:102::4d: Cannot assign requested address

Error: Could not get https://data.zeek.org/traces/ipv6.trace.gz.md5sum, skipping download.
```

Is this expected / "ok" output or is there some issue with Travis at the moment?

**Comment by rsmmr** [comment]
LBNL (which is hosting the data files) is off the grid currently thanks to PG&E cutting power ...

**Comment by Neverlord** [comment]
> LBNL (which is hosting the data files) is off the grid currently thanks to PG&E cutting power ...

Thanks for the quick response. 🙂

Then I'll wait until Travis is fully operational again.

**Comment by rsmmr** [comment]
> Should this get ever voted in: I sincerely hope not under that name. C++ devs have been familiar with this pointer under the name "intrusive pointer" for ages and it has been available in the Boost library as `intrusive_ptr` for well over a decade.

The proposal seems to argue that the semantics of `retain_ptr` and Boost's `intrusive_ptr` are sufficiently different to pick a different name. Do you think that's a valid point? (I don't have an opinion).

> Is it API stability or is the general policy simply not to use CAF headers in Zeek directly?

Both. 

First, I'd want to copy over the header anyways, even if we used the CAF version.  CAF is only an indirect dependency for Zeek itself. For a core class, like this would become, I don't think having that kind of dependency would be great (if CAF ever changed something about that class, it would ripple through Zeek itself, all the plugins, etc.). If we copy it over, we at least control timing for any changes.

Second, as we're introducing a new API here, it seems to make sense to consider functionality that future C++ versions will bring out of the box and go with what's going to be standardized. If we need to copy a header file over to get that now, that seems fine to me (as, per above, we'd likely do that anyways). Now, of course the question is: will `retain_ptr` enter the standard? If yes, I think we should go with it now already. If no, then there's no reason to pick it over other options. From what I'd been reading, `retain_ptr` seemed pretty close to getting in, but maybe I misinterpreted that.

**Comment by Neverlord** [comment]
> Do you think that's a valid point?

No. API details don't matter. A name should transport meaning. And "intrusive pointer" has a well-established meaning, unlike "retain pointer". The difference between `shared_ptr` and `intrusive_ptr` is exactly that the latter requires a reference count in the object itself (or in some other user-provided storage), i.e., *is intrusive*. But that's just my opinion. 😉

Unfortunately the committee has some history on compromising on the name. See `unordered_map`, only because they were afraid `hash_map` might produce some conflicts somewhere.

> I'd want to copy over the header anyways

Ok, then how about this: I use the CAF version as basis for a new `IntrusivePtr` class in Zeek (it's going to be a single, straightforward and dependency-free header) that directly uses the `Ref`/`Unref` API. That way, we don't have to provide any wrapper code now. Should some variant of an intrusive pointer make it into the standard, it's only a small adaption anyway and we can migrate *once* to the final and standardized API. Would that work?

> CAF is only an indirect dependency for Zeek itself.

Noted. I'll keep contributions based on CAF to Broker then. 🙂

> From what I'd been reading, `retain_ptr` seemed pretty close to getting in, but maybe I misinterpreted that.

Hm, I did find trip reports for Kona from Feb '19 that are optimistic that `retain_ptr` should find its way into C++23. So maybe there are enough supporters in the committee at this point to carry it over the next couple of years, but who knows? 🤷🏻‍♂️

File: src/IntrusivePtr.h [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+#include <utility>
+
+// These forward declarations only exist to enable ADL for the Ref and Unref
+// functions.
+
+struct IntrusivePtrDummy;
+
+void Ref(IntrusivePtrDummy*);
+
+void Unref(IntrusivePtrDummy*);
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply src/IntrusivePtr.h]: It will work only if the compiler finds some functions called `Ref` / `Unref` while parsing the template. It probably works for you, because this isn't the first header you include and other includes pull in the function prototypes.

Consider this snippet:

```c++
template <class T>
struct foo {
  int g(T x) {
    return f(x);
  }
};

int f(int x) {
    return x;
}

int main() {
    foo<int> obj;
    return obj.g(42);
}
```

It [fails to compile](https://godbolt.org/z/iT014S), but for a subtle reason.

The call to `f(x)` in `foo::g` depends on `T`, so the compiler will eventually find a fitting overload during ADL. However, the compiler still checks whether a symbol called `f` exists while parsing the template `foo`. The function prototypes here simply prevent the compiler from bailing out early and makes sure it actually does the ADL when instantiating the template.

File: src/IntrusivePtr.h [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+		// nop
+		}
+
+	IntrusivePtr(const IntrusivePtr& other) noexcept
+		{
+		setPtr(other.get(), true);
+		}
+
+	template <class U, class = std::enable_if_t<std::is_convertible_v<U*, T*>>>
+	IntrusivePtr(IntrusivePtr<U> other) noexcept : ptr_(other.detach())
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply src/IntrusivePtr.h]: Sorry, that's an oversight. I'll clean that up and switch to `detach`.

File: src/IntrusivePtr.h [hunk 3]
```diff
@@ -0,0 +1,216 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+// An intrusive, reference counting smart pointer implementation.
```
Comment (rsmmr) [inline src/IntrusivePtr.h]: Mind adding some more API docs for this class? Given that it's now hardcoded to Ref/Unref, it'd be good to describe things from that angle: how is this pointer used to ref'cnt `BroObj` instances? A summary for the class would be good, and some more detail at least on methods that's aren't completely obvious. Basically, make it easy/obvious for people to use this class going forward.

File: src/IntrusivePtr.h [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		if ( raw_ptr && add_ref )
+			Ref(raw_ptr);
+		}
+
+	pointer ptr_;
+};
+
+// Convenience function for creating intrusive pointers.
+template <class T, class... Ts>
+IntrusivePtr<T> makeCounted(Ts&&... args)
```
Comment (rsmmr) [inline src/IntrusivePtr.h]: `makeIntrusive` seems more natural? Or actually even `make_intrusive` for consistency with standard smart pointers.

File: src/broker/Data.cc [hunk 5]
```diff
@@ -227,20 +226,17 @@ struct val_converter {
 
 				if ( ! index_val )
 					{
```
Comment (rsmmr) [inline src/broker/Data.cc]: Can remove the braces now here, and in some more instances in the following.

File: src/IntrusivePtr.h (hunk 4 above)
Comment (Neverlord) [reply src/IntrusivePtr.h]: I don't have a strong preference on the name. What I like about `makeCounted` is that focuses on the *object*: "make a (reference) counted object". Much like the STL functions: `make_shared` for "make a shared object" and `make_unique` for "make a uniquely owned object".

Since we use CamelCase for the class, I think it's more consistent if the factory function also follows that. But there are also many functions in Zeek that use std_case.

So... whichever you guys prefer / is more aligned with the Zeek style. 😉

Comment (timwoj) [reply src/IntrusivePtr.h]: The style is mostly that global functions are snake case and class methods are CamelCase, from what I can tell. I didn't spell that out in the style guide though and it should probably be fixed.

**Comment by Neverlord** [comment]
> As a note about next steps, it could be useful to start adding method overloads that work on intrusive pointers

Full ACK. I only migrated a couple of small functions that aren't broadly used to illustrate how `IntrusivePtr` simplifies the code. I think we can migrate functions that have a limited number of callers directly but should add overloads to functions that have callers throughout the code base to migrate smoothly without doing whole-code-base refactoring. Also, any API function visible to plugins etc. probably has a long migration period.

> More generally, I'm wondering if there's a way to easily convert a bunch of stuff over to receiving IntrusivePtr instead of raw pointers.

Sounds like a job for a custom `clang-tidy` check / autofix. 🙂

I don't have experience with writing such checks, though. So I don't know how much work it'd be.

I agree on your thoughts regarding: conversions should remain explicit. Also, the explicit conversions make it easy to grep for things like `detach` to see where we are still falling back to raw pointers and manual object lifetime management.

**Comment by 0xxon** [comment]
Just to add my 2 cents here - I think this is pretty neat - and much better than the manual counting that we have at the moment.
